### PR TITLE
Add two party ECDSA review spike

### LINKED ISSUE #897: ⚗️ Verify and understand KZen's implementation of two-party ECDSA
KZen has done a rust implementation of Lindell's Two-party ECDSA protocol.
https://github.com/KZen-networks/multi-party-ecdsa

It's GPLv3 which is ok for us since we are too so we may be able to use it.
Going through it and comparing it to the paper will be a good learning experience and will let us know what we will have to do to implement: https://lists.linuxfoundation.org/pipermail/lightning-dev/attachments/20180426/fe978423/attachment-0001.pdf

DoD:
- [x] Spike to summarize learning & findings is merged

### PR DESCRIPTION
**Pull request by LLFourn**
Resolves https://github.com/comit-network/comit-rs/issues/897

### CODE REVIEW

File: 0005-kzen-two-party-ecdsa.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+Issue: [#897](https://github.com/comit-network/comit-rs/issues/897)
+
+## Context
+
+In order to do a [*Scriptless Scripts* atomic swap](https://lists.linuxfoundation.org/pipermail/lightning-dev/attachments/20180426/fe978423/attachment-0001.pdf) with ECDSA signatures we need to have an implementation of Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf) protocol.
+The point of this spike is to review the protocol and KZen's [rust implementation](https://github.com/KZen-networks/multi-party-ecdsa) and make notes about.
+
+## Overview
+
+The paper by Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf)is a way to generate a joint public key and to generate joint signatures under this public key.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
The paper by Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf) is a way to generate a joint public key and to generate joint signatures under this public key.
```

File: 0005-kzen-two-party-ecdsa.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+Here's a list of the main proofs that are needed to implement the protocol.
+
+## Key Generation
+
+### Generating a shared public key
+
+To produce signatures under a shared public key you first need a shared public key.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing like protocol.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
The technique used in the paper (and followed by the implementation) is to do a coin tossing-like protocol.
```

File: 0005-kzen-two-party-ecdsa.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key second.
+
+Note, The ability to simulate a random coin toss gives this protocol very strong security guarantees. Not doing so, doesn't necessarily mean it would be *insecure*.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
Note, The ability to simulate a random coin toss gives this protocol very strong security guarantees.
Not doing so, doesn't necessarily mean it would be *insecure*.
```

I am not really sure, what value the last sentence adds. Does it mean we don't have to do it?

File: 0005-kzen-two-party-ecdsa.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+The implementation commits by sending two hashes, one for the key and one for the NIZK.
+
+The NIZK is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_dlog.rs
+
+### Proving Paillier modulus is well formed
+
+In order to use Paillier encryption P1 needs to generate a RSA type modulus `N = pq` where `p` and `q` are large primes.
+P1 needs to prove that she generated `N` properly (I actually don't know the details of what the risk is to P2 if she doesn't).
+
+The proof mentioned in the original paper is this one: https://eprint.iacr.org/2011/494.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
The proof mentioned in the original paper is this one: [https://eprint.iacr.org/2011/494](https://eprint.iacr.org/2011/494).
```

Comment (thomaseizinger) [reply 0005-kzen-two-party-ecdsa.md]: Instead of embedding the links, should we keep references at the bottom?

File: 0005-kzen-two-party-ecdsa.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+Roughly speaking, the strategy is to prove that `N` is *square free* and to show that `f(x) = x ^ N (mod N)` is a permutation which proves that `N = pq`.
+
+#### Implementation notes
+
+Since it's a public coin protocol, you can make an non-interactive version of the proof using the [Fiat-Shamir transformation](https://en.wikipedia.org/wiki/Fiat%E2%80%93Shamir_heuristic).
+They've implemented a non-interactive version of it here: https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/correct_key_ni.rs (which uses the interactive proof module but supplies the challenge as a hash of the initial commitment).
+
+### Proving a Paillier encryption is a private key of a EC public key
+
+In order to produce two party aggregated ECDSA signatures on a joint Secp256k1 public key we need one party (P1) their Paillier encrypted private key over to P2.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
In order to produce two party aggregated ECDSA signatures on a joint Secp256k1 public key we need one party (P1) to send their Paillier encrypted private key over to P2.
```
?

File: 0005-kzen-two-party-ecdsa.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+But in order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof for this was developed in paper. It's a little tricky to get your head around.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]: :x: Missing reference to paper?

Also:
```suggestion
The proof for this was developed in paper.
It's a little tricky to get your head around.
```

File: 0005-kzen-two-party-ecdsa.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+It works by the P2 doing some homomorphic operations on the encrypted private key and sending the result to P1.
+P1 then decrypts it and figures out the corresponding public key.
+
+The weird thing is that P1 doesn't actually use his knowledge of what the private key is in order to produce the proof that the cipher text doesn't contain the private key.
+He simply uses his ability to decrypt the Paillier private key!
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]: I am not sure if I get this.
If it is important, it would be good to elaborate on this a bit more.

File: 0005-kzen-two-party-ecdsa.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/party_one.rs
+
+It looks like this protocol cannot be made non-interactive. I think it's the only one that couldn't be so it accounts for most of the rounds of communication.
+It has four rounds of communication.
+
+### Range proof
+
+In order for the previous proof to actually prove the statement you have to couple it with a range proof which proves that the encrypted private key is in the curve order (i.e. is a valid private key).
+The poof chosen was originally from [https://www.iacr.org/archive/eurocrypt2000/1807/18070437-new.pdf](https://www.iacr.org/archive/eurocrypt2000/1807/18070437-new.pdf) but I found it was easier to understand in Lindell's paper anyway (see Appendix A).
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]: Click-able references (for the "Lindell's paper" part) would be really good here (plus IEEE citation and a list in the end).

File: 0005-kzen-two-party-ecdsa.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+The proof uses the cut and choose technique so it's quite large.
+It's tricky to understand, but doesn't use any special math. You just have to follow what happens closely.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]:
```suggestion
It's tricky to understand, but doesn't use any special math.
You just have to follow what happens closely.
```

File: 0005-kzen-two-party-ecdsa.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+It's implemented here:
+
+https://github.com/KZen-networks/zk-paillier/blob/a7f3907fb2b3464f12b55e3ad2f50405db54f36a/src/zkproofs/range_proof.rs
+
+
+## Messages
+
+### Keygen
+
+Here's my early sketch of how many messages you need:
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.md]: Nice, thanks for including that!

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Really good work! :)

Most comments are formatting-related.
A references section would be really good in order to keep track of when you are referring to which paper. Seeing [1] and [2] is (at least for me) easier to follow than names and links.

File: 0005-kzen-two-party-ecdsa.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a [Paillier group](https://en.wikipedia.org/wiki/Paillier_cryptosystem) rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
+
+Once the keys are generated, Paillier is used to sign roughly in the this way:
+
+1. P1 sends the Paillier encryption of their private key to P2 (this is actually done in the keygen phase)
+2. P2 uses Paillier homorphic operations with the encrypted private key (addition and multiplication) to build a new ciphertext which has P2's parts of the signature in it and sends it to P1.
+3. P1 then decrypts it and then pretty much has a valid ECDSA signature.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.md]: why just `pretty much` a valid signature?

File: 0005-kzen-two-party-ecdsa.md (hunk 7 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.md]: Heh that doesn't should be a "does". But yes it is hard to explain. I'll see what i can do.

File: 0005-kzen-two-party-ecdsa.adoc [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+
+== Context
+
+In order to do a _Scriptless Scripts_ atomic swap <<Sanchez18a>> with ECDSA signatures we need to have an implementation of <<Lindell17>>.
+The point of this spike is to review <<Lindell17>> and KZen's rust implementation <<kzen_mp_ecdsa>> and make notes about.
+
+
+== Overview
+
+[[Lindell17]] describes a two-party protocol to (i) generate a joint public key and (ii) jointly create ECDSA signatures under that public key.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: I assume this is the correct formatting as otherwise nothing is rendered. 
```suggestion
<<Lindell17>> describes a two-party protocol to (i) generate a joint public key and (ii) jointly create ECDSA signatures under that public key.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`). There's no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
+
+Since ECDSA signatures doesn't have a nice algebraic structure it is **very** difficult to do signature aggregation.
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a [Paillier group](https://en.wikipedia.org/wiki/Paillier_cryptosystem) rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: 😅

File: 0005-kzen-two-party-ecdsa.adoc [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+For us it may be very useful. Even though Bitcoin will get Schnorr signatures, Ethereum probably won't switch anytime soon.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: 🙃
```suggestion
For us it may be very useful. 
Even though Bitcoin will get Schnorr signatures, Ethereum probably won't switch anytime soon.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+
+The hard bit is doing the setup phase which involves quite a few proofs.
+Some of them are in the above crate but some of them are imported from elsewhere.
+We'll start with going through these.
+
+== Key Generation
+
+=== Generating the joint public key
+
+To produce signatures under a joint public key you first need generate it fairly.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
To produce signatures under a joint public key you first need to generate it fairly.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+
+To do the signing protocol, we need one party (P2) to have the encrypted pivate key of the other (P1).
+In order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof is described in <<Lindell17>> (section 6).
+It's a little tricky to get your head around.
+First P2 sends a challenge to P1.
+If P1's response to the challenge is valid then P2 knows that the encrypted private key corresponds to `p1_public_key`.
+
+The weird thing is that P1 doesn't actually use his knowledge of what private key is in order to produce the proof the encrypted private key is correct.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: Can you rephrase this sentence please? I'm having a hard time to understand it, especially this part here: `use his knowledge of what private key is in order to produce the proof `.

File: 0005-kzen-two-party-ecdsa.adoc [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+The proof is not showing knowledge of the private key, just that the encrypted private key is indeed the preimage of `p1_public_key`.
+
+==== Implementation notes
+
+This proof isn't in its own module but the code for it hangs around with the rest of the protocol:
+
+https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/party_one.rs
+(look for things starting with `PDL`).
+
+It looks like this protocol cannot be made non-interactive. It requires four round of communication.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
It looks like this protocol cannot be made non-interactive. It requires four rounds of communication.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 18]
```diff
@@ -0,0 +1,173 @@
...
+* Authors(s): {authors}
+* Date: {revdate}
+
+Issue: https://github.com/comit-network/comit-rs/issues/897[#897]
+
+
+== Context
+
+In order to do a _Scriptless Scripts_ atomic swap <<Sanchez18a>> with ECDSA signatures we need to have an implementation of <<Lindell17>>.
+The point of this spike is to review <<Lindell17>> and KZen's rust implementation <<kzen_mp_ecdsa>> and make notes about.
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: I'm missing some kind of conclusion, e.g. is the provided library sufficient for us to be used right away?

File: 0005-kzen-two-party-ecdsa.adoc (hunk 16 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: I reworded this section a little bit (added a "that" to the sentence).

File: 0005-kzen-two-party-ecdsa.adoc [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+
+
+== Overview
+
+<<Lindell17>> describes a two-party protocol to (i) generate a joint public key and (ii) jointly create ECDSA signatures under that public key.
+This is usually called signature _aggregation_.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]: :upside_down_face: `usually`?
```suggestion
This is called signature _aggregation_.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+There's a video by the Lindell summarizing his paper here: https://www.youtube.com/watch?v=pwc_Ork-1aA
+
+Aggregated signatures are a nice thing to have.
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`). There's no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`).
There's no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+Since ECDSA signatures doesn't have a nice algebraic structure it is **very** difficult to do signature aggregation.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
Since ECDSA signatures don't have a nice algebraic structure it is **very** difficult to do signature aggregation.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees (they are able to do a simulation based proof).
+
+The protocol is roughly as follows:
+
+- P1 -> Hash(p1_public_key, schnorr_nizk(p1_public_key))
+- P2 -> p2_public_key, schnorr_nizk(p2_public_key)
+- P1 -> Opens commitment from first message
+- Both parties calculate `joint_public_key = p1_public_key + p2_public_key`.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]: I guess before that, both parties have to actually verify the ZKP?

File: 0005-kzen-two-party-ecdsa.adoc [hunk 23]
```diff
@@ -0,0 +1,173 @@
...
+
+
+=== Proving a Paillier encryption is a private key of a EC public key
+
+To do the signing protocol, we need one party (P2) to have the encrypted pivate key of the other (P1).
+In order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof is described in <<Lindell17>> (section 6).
+It's a little tricky to get your head around.
+First P2 sends a challenge to P1.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]: The challenge in this case means "sign this thing for me and I will verify against your PK"?

File: 0005-kzen-two-party-ecdsa.adoc [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+If P1's response to the challenge is valid then P2 knows that the encrypted private key corresponds to `p1_public_key`.
+
+The weird thing is that P1 doesn't actually use his knowledge of what private key is in order to produce the proof that the encrypted private key is correct.
+Anyone who knows the factorization `N = pq` can make this proof to P2.
+This is becasuse the proof is not showing knowledge of the private key, just that the encrypted private key is indeed the preimage of `p1_public_key`.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]: Please spend some time in the presentation on this, I still don't get it.

File: 0005-kzen-two-party-ecdsa.adoc [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+== Signing
+
+Assuming the keygen phase went well we have the parties knowing the following (where `x1`,`x2` are private keys and `X` is the joint public key).
+
+1. P1 knows: `x1` , `X`, `(N,p,q | N = pq)`,
+2. P2 knows: `x2`, `X`, `N`, `c1 = PaillierEncrypt(x1, N)`
+
+Now they want to sign a message `m`.
+
+Since ECDSA signatures are in the form `(r,s)`.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]: :question: English question. Starting this sentence with `Since` makes me want to combine it with the 2nd one.
Right or wrong?

**Review by thomaseizinger — COMMENTED** [review]
Good work :)

I left a few suggestions + asked a few questions!

File: 0005-kzen-two-party-ecdsa.adoc (hunk 18 above)
Comment (thomaseizinger) [reply 0005-kzen-two-party-ecdsa.adoc]: To me it sounded like it is :) 

But I agree, a conclusion clearly stating that would be good!

File: 0005-kzen-two-party-ecdsa.adoc (hunk 22 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: Yes they certainly must do that and p2 must verifying the opening of the commitment. I those steps out though because they seem obvious.

File: 0005-kzen-two-party-ecdsa.md (hunk 3 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.md]: Yeah I removed it. We might not have to do it. AFAIK as long as you protect against rouge key attacks with the NIZK proof you should be fine.

i.e. You could do something like this in the *Knowledge of Secret Key* model  (KOSK) as far as I understand.
- P1 -> sends `p1_public_key`,
- P2 -> sends `p2_public_key, schnorr_nizk(p2_public_key)`

The 3 round coin tossing is there because Lindell is doing a simulation based proof. In particular he is simulating a functionality which chooses a uniformly random group element. So the extra messages are an artifact of the proof technique (at least to some extent).

I'm not an expert on this so of course we should just do what the paper says :)

File: 0005-kzen-two-party-ecdsa.md (hunk 11 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.md]: Because P1 will still have to do some operations on the decrypted value to actually produce the valid signature.

Note I moved and expanded this section to the end.

File: 0005-kzen-two-party-ecdsa.adoc (hunk 18 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: I added an `== Outcome Summary` at the beginning.

File: 0005-kzen-two-party-ecdsa.adoc (hunk 19 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: Actually the term is wrong. It's called "a signature scheme with key aggregation".

(see https://eprint.iacr.org/2018/068.pdf)

File: 0005-kzen-two-party-ecdsa.adoc (hunk 23 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: No. It was probably a mistake to try and explain the protocol here. I expanded it so it at least provides some value but it's best to just try and read the paper.

File: 0005-kzen-two-party-ecdsa.adoc (hunk 25 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: Yes. A mistaken end of sentence.

File: 0005-kzen-two-party-ecdsa.adoc [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing protocol to determine the joint public key.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees (they are able to do a simulation based proof).
+
+The protocol is roughly as follows, where x~1~ is P1's (resp. P2) private key and X~1~ = g^x~1~^  is her public key.
+
+- P1 -> Hash(X~1~, schnorr_nizk(X~1~))`
```
Comment (da-kami) [inline 0005-kzen-two-party-ecdsa.adoc]: 🔧Like this it would be more visible that they actually exchange something. :
```
- P1 -> P2: Hash(X~1~, schnorr_nizk(X~1~))
- P2 -> P1: X~2~, schnorr_nizk(X~2~)
- P1: Opens commitment from first message
- Both parties calculate the joint public key X = X~1~^x2^ = X~2~^x1^ (i.e. Diffie-Hellman)
```
However, it might be just my brain that sees it like that - feel free to leave it as it is if it is understandable to others or this notation conforms to some standard :)

File: 0005-kzen-two-party-ecdsa.adoc [hunk 27]
```diff
@@ -0,0 +1,182 @@
...
+To produce signatures under a joint public key you first need to generate it fairly.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing protocol to determine the joint public key.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees.
+
+The protocol is roughly as follows, where x~1~ is P1's (resp. P2) private key and X~1~ = g^x~1~^  is her public key.
+
+- P1 -> Hash(X~1~, schnorr_nizk(X~1~))`
```
Comment (da-kami) [inline 0005-kzen-two-party-ecdsa.adoc]: 🔧  remove the additional `

File: 0005-kzen-two-party-ecdsa.adoc [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+In order for P2 to know that he's signing under their joint public key, he has to know that this encrypted private key is the right one (without learning what the private key is).
+
+The proof is described in <<Lindell17>> (section 6).
+It's a little tricky to get your head around.
+Here's a high level and very rough intuition to how it works (it would be better to read the paper):
+
+P2 takes the P1's encrypted private key c~key~ and multiplies and adds random stuff to it (using Paillier operations).
+This creates a new encrypted private key c~key~'.
+P2 has no idea what about the value of this new private key he's just created **but** he can figure out the corresponding public key because he knows P1's public key and the values he added to c~key~.
+P2 sends c~key~'` to P1 (think of it as his challenge to P1).
```
Comment (da-kami) [inline 0005-kzen-two-party-ecdsa.adoc]: 🔧  remove the additional `

File: 0005-kzen-two-party-ecdsa.adoc [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+
+P1 decrypts c~key~` to this random private key and figures out the public key.
```
Comment (da-kami) [inline 0005-kzen-two-party-ecdsa.adoc]: 🔧  I think it should be ' instead of `

**Review by da-kami — APPROVED** [review]
Really nice work and great summary of complicated topics :)

File: 0005-kzen-two-party-ecdsa.adoc [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+
+==== Implementation notes
+The implementation commits by sending two hashes, one for the key and one for the NIZK.
+
+The NIZK is implemented here: https://github.com/KZen-networks/curv/blob/18f0081a9a3025eafd49ad496e5032647edf4aa3/src/cryptographic_primitives/proofs/sigma_dlog.rs
+
+=== Proving Paillier modulus is well formed
+
+In order to use Paillier encryption P1 needs to generate a RSA type modulus `N = pq` where `p` and `q` are large primes.
+P1 needs to prove that she generated `N` properly (I don't actually know what the risk is to P2 if she doesn't).
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: According to the paper this is needed in order to proof that `gcd(N, φ(N) = 1` holds and hence is indeed additively homomorphic.

File: 0005-kzen-two-party-ecdsa.adoc [hunk 31]
```diff
@@ -0,0 +1,182 @@
...
+P1 then decrypts c~3~ and does a modular reduction to the curve order (this transforms it from a scalar in the Paillier group to a scalar in the elliptic curve group).
+From this, P~1~ can produce `s` and therefore a valid `(r,s)` ECDSA signature on `m`.
+
+==== Implementation Notes
+
+The code for the signing part is in the main 2pECDSA crate: https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/
+
+=== Messages
+
+The messages are nicely pictures in section 3.3 figure 1
```
Comment (bonomat) [inline 0005-kzen-two-party-ecdsa.adoc]: Section 3.3. Figure 1 of what paper? :)

**Review by bonomat — COMMENTED** [review]
Good work on that, before you merge it I'm happy to do a last cleanup as there are some minor changes I'd like to do here and there (I don't want to clutter the PR with typo/style corrections).

**Comment by D4nte** [comment]
Trying a quick read in the train but I think I need a proper isolated time to read that, will try tomorrow.
It would be interesting to have Pedro's input on this document and see if he has any comments :)

File: 0005-kzen-two-party-ecdsa.adoc (hunk 30 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: But what does this mean for P2? Does P2 leak secret information somewhere if it's not? Or maybe P2 will accept an invalid proof from P1 if this doesn't hold. I'm sure there's a reason but didn't come across it the paper (yet!).

File: 0005-kzen-two-party-ecdsa.adoc (hunk 26 above)
Comment (LLFourn) [reply 0005-kzen-two-party-ecdsa.adoc]: I did this :)

File: 0005-kzen-two-party-ecdsa.adoc [hunk 32]
```diff
@@ -0,0 +1,182 @@
...
+<<Lindell17>> describes a two-party protocol to (i) generate a joint public key and (ii) jointly create ECDSA signatures under that public key.
+The technical term for this is _key aggregation_.
+Lindell summarizes his paper in the following https://www.youtube.com/watch?v=pwc_Ork-1aA[video].
+
+Signature schemes with aggregated keys are a nice thing to have.
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you do not need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`).
+There is no way to tell the difference between a single user `P2WSH` output and a `P2WSH` output from a joint public key.
+
+Since ECDSA signatures don not have a nice algebraic structure it is **very** difficult to do key aggregation.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
Since ECDSA signatures do not have a nice algebraic structure it is **very** difficult to do key aggregation.
```

File: 0005-kzen-two-party-ecdsa.adoc [hunk 33]
```diff
@@ -0,0 +1,182 @@
...
+== Key Generation
+
+=== Generating the joint public key
+
+To produce signatures under a joint public key you first need to generate it fairly.
+Because of rogue key attacks, you can not just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing protocol to determine the joint public key.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key afterwards.
+Note, The ability to simulate a random coin toss gives the key generation very strong security guarantees.
```
Comment (thomaseizinger) [inline 0005-kzen-two-party-ecdsa.adoc]:
```suggestion
Note: The ability to simulate a random coin toss gives the key generation very strong security guarantees.
```

**Review by thomaseizinger — APPROVED** [review]
